List audio devices on Windows

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

Add listing devices with PortAudio backend (by default on Windows). With this PR following commands will work on Windows:
```supercollider
ServerOptions.inDevices;
ServerOptions.outDevices;
ServerOptions.devices;
```

Fixes #1557.

RE #1557 - the issue mentions listing devices not working on Linux as well, but since we use JACK on Linux there's [no point implementing listing devices](https://github.com/supercollider/supercollider/issues/1557#issuecomment-114677077). (Maybe at some point we could consider implementing available JACK server names... but this PR is really for Windows users and PortAudio).

## Types of changes

<!-- Delete lines that don't apply -->

- Documentation
- New feature

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- ~~All tests are passing~~
  - there are no tests
- [x] Updated documentation
- [x] This PR is ready for review
- [x] Cleanup commit history

brianlheim commented:
does this depend on another PR? i see a bunch of CMake changes that aren't really noted in the PR description

brianlheim commented:
oh, i see now, it's because sclang now needs to link with PortAudio

brianlheim commented on lang/LangPrimSource/PyrPrimitive.cpp:
```diff
@@ -4348,11 +4348,16 @@ void initPrimitives() {
+    void initPortAudioPrimitives();
+    initPortAudioPrimitives();
+#endif
+
```
these two paragraphs would become

```cpp
#if defined(SC_AUDIO_API_COREAUDIO) || defined(SC_AUDIO_API_PORTAUDIO)
    initAudioBackendPrimitives();
#endif
```

dyfer replied:
> so, i'm thinking that to avoid code duplication it's probably best to merge `SC_CoreAudioPrim.cpp` and `SC_PortAudioPrim.cpp` into a new C++ file, `SC_AudioBackendPrim.cpp` (feel free to choose a better name, i'm not sure that one is accurate).

Below you also suggest refactoring what I wrote for PortAudio part. How would you like the combined file structured? Should CoreAudio part be also refactored?

brianlheim replied:
no, the CoreAudio part doesn't need to be refactored, you can just stick a ifdef around it. if i've read the code right, then prListAudioDevices and initXPrimitives won't be duplicatedlike they are now

brianlheim commented on lang/LangPrimSource/SC_PortAudioPrim.cpp:
```diff
@@ -0,0 +1,113 @@
+            g->gc->GCWriteNew(devArray, (PyrObject*)string); // we know array is white so we can use GCWriteNew
+            j++;
+        }
+    }
```
this function body could be written much more idiomatically. essentially, it's 3 steps:
1. collect device infos (which already contain the name of the device, btw; you don't need `GetPaDeviceName` here)
2. filter by the IN/OUT condition
3. transform filtered device infos into a PyrArray

the first step is just a `for` loop filling a `std::vector`; the second step you can write with the standard library functions [std::remove_if](https://en.cppreference.com/w/cpp/algorithm/remove) and [std::vector::erase](https://en.cppreference.com/w/cpp/container/vector/erase). you can use the size of the vector after filtering as a replacement for the variable `num` in this function body.

the third step is then also a simple `for` loop.

since device names are null-terminated `const char*`, you don't have to use an intermediary `std::string` to hold that data.

dyfer replied:
OK, I'll try to implement the the logic you suggest.
However, we do need `GetPaDeviceName` - on Windows that returns `backend name : device name` and is consistent with what the server uses.

brianlheim replied:
ah okay, i forgot it does more than just read the name field on Windows. probably worth adding an overloading taking a PaDeviceInfo so you don't need to call PaGetDeviceInfo twice and keep around a list of indices

brianlheim replied:
an overload of GetPaDeviceName, i mean

dyfer replied:
Do you mean for GetPaDeviceName to be able to take device info on top of device index?

brianlheim replied:
(answered on slack)

dyfer commented:
@brianlheim I think I addressed your suggestions. How does the PortAudio section look like now?

I consciously made the change to always add the device selection file (regardless of the API), also as an attempt to make CMakeLists a little less cluttered _(very little...)_. I think the only downside is that it does change the type of error (when calling `ServerOptions.devices`) when compiled with JACK/on Linux. I don't think this is a big issue but I wanted to point it out.

dyfer commented on lang/LangPrimSource/SC_AudioDevicePrim.cpp:
```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
+    if (numDevices < 0)
+        return 0;
+
+    std::vector<const PaDeviceInfo*> deviceInfos;
```
@brianlheim do I need to deallocate this vector?

brianlheim replied:
nope, that is the magic of [RAII](https://en.cppreference.com/w/cpp/language/raii) -- in short, much of modern C++ relies on the concept of treating an object's constructor as an initialization and resource acquisition step, and treating its destructor (here, called when we exit the function) as a resource freeing step. opening files (std::fstream), allocating memory (std::vector and other std containers), creating threads (std::thread), and acquiring/releasing a lock (std::lock_guard) all can work this way in C++.

brianlheim commented:
> I think the only downside is that it does change the type of error (when calling ServerOptions.devices) when compiled with JACK/on Linux. I don't think this is a big issue but I wanted to point it out.

that seems fine to me, as long as it fails; it's documented as not being implemented anyway.

brianlheim commented on lang/LangPrimSource/SC_AudioDevicePrim.cpp:
```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
+                    deviceInfos.push_back(pdi);
+            }
+        } else
+            deviceInfos.push_back(pdi);
```
i think you kind of copied this from above, but it's a really awkward way to write it. we know `type` can only have 3 values, so a series of if-elses is a clearer way to write it:

```cpp
if (type == OUT) {
    ...
} else if (type == IN) {
    ...
} else {
    // type is BOTH
    ...
}
```

brianlheim commented on lang/LangPrimSource/SC_AudioDevicePrim.cpp:
```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
+
+    PyrSlot* a = g->sp - 2;
+    PyrObject* devArray = newPyrArray(g->gc, numSelectedDevices * sizeof(PyrObject), 0, true);
+    SetObject(a, devArray); // this is okay here as we don't use the receiver below
```
having a local variable named `a` is not really helping here -- you can just write `SetObject(g->sp - 2, devArray)`

brianlheim commented on lang/LangPrimSource/SC_AudioDevicePrim.cpp:
```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
+    for (int i = 0; i < numSelectedDevices; i++) {
+        PyrString* string = newPyrString(g->gc, GetPaDeviceName(deviceInfos[i]).c_str(), 0, true);
+        SetObject(devArray->slots + i, string);
+        devArray->size++;
```
you can simply set `devArray->size = numSelectedDevices` above or below this loop, this is kind of unnecessary work

brianlheim commented on lang/LangPrimSource/SC_AudioDevicePrim.cpp:
```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
+            deviceInfos.push_back(pdi);
+    }
+
+    int numSelectedDevices = deviceInfos.size();
```
personally, i would just use `deviceInfos.size()` instead of creating another local variable. it's the same number of characters, and typically having fewer local variables (within reason) makes code easier to understand. when you see `numSelectedDevices` below, it's not immediately obvious that it has a relationship to `deviceInfos`, and it's definitely not obvious that that relationship stays intact. small point, but just wanted to state my preference

brianlheim commented on lang/LangPrimSource/SC_AudioDevicePrim.cpp:
```diff
@@ -125,6 +131,56 @@ int listDevices(struct VMGlobals* g, int type) {
+        PyrString* string = newPyrString(g->gc, GetPaDeviceName(deviceInfos[i]).c_str(), 0, true);
+        SetObject(devArray->slots + i, string);
+        devArray->size++;
+        g->gc->GCWriteNew(devArray, (PyrObject*)string); // we know array is white so we can use GCWriteNew
```
generally try to avoid c-style casts when a less aggressive cast will work -- c-style casts can do lots of conversion including changing const-ness (https://en.cppreference.com/w/cpp/language/explicit_cast). here, you should use `static_cast<PyrObject*>(string)`

dyfer replied:
I can't say that I understand all these particular objects (`PyrString`, `PyrObject`) and this was copied from the coreaudio code. However, when I tried to use your suggestion, I got 
```
error: static_cast
      from 'PyrString *' to 'PyrObject *', which are not related by inheritance, is not allowed
```

brianlheim replied:
ah, my mistake, i thought PyrString was a derived class of PyrObject, but it's not. you'll have to use `reinterpret_cast`.

dyfer commented on HelpSource/Reference/AudioDeviceSelection.schelp:
```diff
@@ -24,7 +24,7 @@ code::
 ::
 
-On macOS you can programmatically obtain a list of available audio devices without booting the server:
+You can programmatically obtain a list of available audio devices without booting the server:
```
Sounds good. I went with your first suggestion, since it seems more user-friendly. It the user compiles SC with a different backend, they likely understand the implications of different backends.

brianlheim commented:
thanks!
